docs/proposals/generic-web-api-provider-config.md

```diff
@@ -0,0 +1,145 @@
+
+### Detailed examples
+
+#### Example 1: Authenticating to Upbound
+
+[Upbound][upbound] uses 60 minute short-lived tokens to authenticate users to access their managed control planes using their robot tokens (API keys). In order to exchange their robot token, users must make a POST request to the `/orgscopedtokens` API passing a form body that includes the name of the Upbound organization to authenticate against and the robot token. If successful, this API returns a JSON payload which includes a string `access_token` to be used as the bearer token when communicating with the Upbound control plane.
+
+```yaml!
+webAPIProviderConfig:
+  method: "POST"
+  url: https://auth.upbound.io/apis/tokenexchange.upbound.io/v1alpha1/orgscopedtokens
+  # this example uses ${.orgName} and {.token}, which are provided by the user as part of the secret
+  body: "audience=upbound%3Aspaces%3Aapi&audience=upbound%3Aspaces%3Acontrolplanes&grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Atoken-exchange&scope=upbound%3Aorg%3A${.orgName}&subject_token={.token}&subject_token_type=urn%3Aietf%3Aparams%3Aoauth%3Atoken-type%3Aid_token"
+  headers:
+    "Content-Type": "application/x-www-form-urlencoded"
+  tokenPath: $.access_token
+```
+
+[upbound]: https://www.upbound.io/
+
+### Security Considerations
+
+This proposal suggests that Argo CD make calls to arbitrary endpoints. Ideally the user would be able to validate the security of the endpoint independently before configuring Argo CD to interact with it.
+
+Argo CD will require the use of TLS in its requests to the external APIs to protect against man-in-the-middle attacks that could compromise users' credentials during transmission.
+
+Argo CD will only store the credentials returned by the APIs in memory and will never persist them to a volume or secret that may be accessible by other users.
+
+### Risks and Mitigations
+
+
+
+### Upgrade / Downgrade Strategy
+
+No action is needed by the user to upgrade to the latest version of Argo CD that supports generic web API provider configurations. The feature is entirely opt-in and users that do not choose to opt-in will see no change.
+
+When downgrading, users that have opted-in to using the configuration will not be able to access clusters that are using it to authenticate. Users will either need to revert to long-lived credentials or will need to find workaround solutions to authenticating these clusters using short-lived credentials (such as the ones mentioned in the "Motivation" section).
+
+## Drawbacks
+
+Asking users to provide the parameters for an authentication request is a poor user experience. Users may not be aware of the authentication APIs for their cluster providers and therefore might not know what values are valid for their Argo CD cluster. This poor experience may be offset by documentation, provided either by Argo and/or by the cluster providers, with examples of how to configure the requests.
+
+## Alternatives
```

Comment:
Would using Dex's builtin AuthProxy method work/be a valuable alternative here? https://dexidp.io/docs/connectors/authproxy/

Reply:
I'm not familiar with Dex in this context. Would this require any user to configure Dex as the authentication platform for Argo?